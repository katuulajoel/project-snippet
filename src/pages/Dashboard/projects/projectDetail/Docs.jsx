/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React, { useState } from "react";
import _ from "lodash";
import { useDispatch, useSelector } from "react-redux";

/* -------------------------- Internel Dependencies ------------------------- */

import {
  DOC_TYPE_ESTIMATE,
  DOC_TYPE_OTHER,
  DOC_TYPE_PROPOSAL,
  DOC_TYPE_REQUIREMENTS,
  DOCUMENT_TYPES_MAP,
  DOC_TYPE_CONTRACT,
} from "../../../../configs/constants/projectConstants";

/* --------------------------- Component Dependencies -------------------------- */

import {
  createDocument,
  deleteDocument,
} from "../../../../redux/actions/ProjectActions";
import {
  isAdminOrClient,
  isDev,
  isPMAndHasProjectAcess,
} from "../../../../utils/auth";
import Icon from "../../../../components/Icon";
import IconButton from "../../../../components/IconButton";
import DocumentPicker from "../../../../components/DocumentPicker";

const Docs = ({ project }) => {
  const dispatch = useDispatch();
  const { documents: projectDocuments } = useSelector(
    ({ Projects }) => Projects
  );

  const [documents, setDocuments] = useState({
    estimate: [],
    proposal: [],
    requirements: [],
    contract: [],
    other: [],
  });

  const onChangeDocs = (key, docs) => {
    if (docs.length > documents[key].length) {
      let newDoc = docs[docs.length - 1];
      dispatch(
        createDocument({
          ...newDoc,
          project: { id: project.id },
        })
      );
    }

    let newState = {};
    newState[key] = docs;
    setDocuments({ ...documents, ...newState });
  };

  const onRemoveDoc = (id) => {
    dispatch(deleteDocument(id));
  };

  return (
    <div className="project-docs">
      {[
        ...(isDev() ? [] : [DOC_TYPE_PROPOSAL]),
        DOC_TYPE_REQUIREMENTS,
        ...(isDev() ? [] : [DOC_TYPE_ESTIMATE]),
        DOC_TYPE_CONTRACT,
        DOC_TYPE_OTHER,
      ].map((docType) => {
        return (
          <div key={`doc-type-${docType}`} className="section">
            <div className="section-title">
              <span>
                {(DOCUMENT_TYPES_MAP[docType] || _.upperFirst(docType)).replace(
                  /\s?document/,
                  ""
                )}{" "}
                {!(docType === DOC_TYPE_CONTRACT) ? "Documents" : null}
              </span>
              {(isAdminOrClient() || isPMAndHasProjectAcess(project)) &&
                !project.archived && (
                  <>
                    <DocumentPicker
                      showSelected={false}
                      documentType={docType}
                      onChange={(docs) => {
                        onChangeDocs(docType, docs);
                      }}
                    />{" "}
                  </>
                )}
            </div>

            <div
              className={`file-list ${
                !isDev() && !project.archived ? "" : "readonly"
              }`}
            >
              {projectDocuments
                .filter((doc) => doc.type === docType)
                .map((doc) => {
                  return (
                    <div
                      className="section-item"
                      key={`doc-${doc.id}`}
                      style={{ marginBottom: "16px" }}
                    >
                      <a
                        href={doc.download_url}
                        target="_blank"
                        rel="noreferrer"
                        style={{ textDecoration: "auto" }}
                      >
                        <Icon
                          name={
                            doc.file ? "file-document-outline" : "link-variant"
                          }
                        />{" "}
                        {doc.title || "Document Name"}
                      </a>
                      {(isAdminOrClient() || isPMAndHasProjectAcess(project)) &&
                        !project.archived && (
                          <IconButton
                            name="delete-outline"
                            size="main"
                            className="btn-edit"
                            onClick={() => onRemoveDoc(doc.id)}
                          />
                        )}
                    </div>
                  );
                })}
            </div>
          </div>
        );
      })}
    </div>
  );
};

Docs.propTypes = {
  project: PropTypes.object,
  ProjectActions: PropTypes.object,
};

export default Docs;
