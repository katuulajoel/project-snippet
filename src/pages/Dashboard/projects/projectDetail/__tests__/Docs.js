import React from "react";
import renderer from "react-test-renderer";
import Docs from "../Docs";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
// import { mount } from "enzyme/build";

const middlewares = [thunk];

const mockAppState = {
  Invoice: {
    isMakingRequest: {},
    errors: {},
    summary: {},
    list: { data: [], count: 0, next: "", previous: "" },
    invoice: {},
    csv: {},
  },
  Projects: {
    isMakingRequest: {},
    project: null,
    documents: [{ type: "other" }],
  },
};

const mockAppStore = (state) => {
  const mockStore = configureStore(middlewares);
  return mockStore(state);
};

describe("Project Payments test", () => {
  it("Should match snapshot test", () => {
    global.URL.createObjectURL = jest.fn(() => "details");
    const tree = renderer
      .create(
        <Provider store={mockAppStore(mockAppState)}>
          <Docs project={{ id: 123 }} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
