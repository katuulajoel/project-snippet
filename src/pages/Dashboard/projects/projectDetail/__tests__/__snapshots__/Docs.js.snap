// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Payments test Should match snapshot test 1`] = `
<div
  className="project-docs"
>
  <div
    className="section"
  >
    <div
      className="section-title"
    >
      <span>
        Proposal
         
        Documents
      </span>
    </div>
    <div
      className="file-list "
    />
  </div>
  <div
    className="section"
  >
    <div
      className="section-title"
    >
      <span>
        Requirements
         
        Documents
      </span>
    </div>
    <div
      className="file-list "
    />
  </div>
  <div
    className="section"
  >
    <div
      className="section-title"
    >
      <span>
        Estimate
         
        Documents
      </span>
    </div>
    <div
      className="file-list "
    />
  </div>
  <div
    className="section"
  >
    <div
      className="section-title"
    >
      <span>
        Contracts
         
      </span>
    </div>
    <div
      className="file-list "
    />
  </div>
  <div
    className="section"
  >
    <div
      className="section-title"
    >
      <span>
        Other
         
        Documents
      </span>
    </div>
    <div
      className="file-list "
    >
      <div
        className="section-item"
        style={
          Object {
            "marginBottom": "16px",
          }
        }
      >
        <a
          rel="noreferrer"
          style={
            Object {
              "textDecoration": "auto",
            }
          }
          target="_blank"
        >
          <i
            className="tg-ic-link-variant  "
          />
           
          Document Name
        </a>
      </div>
    </div>
  </div>
</div>
`;
